sweepbatcher/greedy_batch_selection.go

```diff
@@ -0,0 +1,315 @@
+package sweepbatcher
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/btcsuite/btcd/txscript"
+	sweeppkg "github.com/lightninglabs/loop/sweep"
+	"github.com/lightningnetwork/lnd/input"
+	"github.com/lightningnetwork/lnd/lntypes"
+	"github.com/lightningnetwork/lnd/lnwallet/chainfee"
+)
+
+// greedyAddSweep selects a batch for the sweep using the greedy algorithm,
+// which minimizes costs, and adds the sweep to the batch. To accomplish this,
+// it first collects fee details about the sweep being added, about a potential
+// new batch composed of this sweep only, and about all existing batches. Then
+// it passes the data to selectBatch() function, which emulates adding the sweep
+// to each batch and creating new batch for the sweep, and calculates the costs
+// of each alternative. Based on the estimates of selectBatch(), this method
+// adds the sweep to the batch that results in the least overall fee increase,
+// or creates new batch for it. If this method fails for whatever reason, the
+// caller falls back to the simple algorithm (method handleSweep).
+func (b *Batcher) greedyAddSweep(ctx context.Context, sweep *sweep) error {
+	if b.customFeeRate == nil {
+		return errors.New("greedy batch selection algorithm requires " +
+			"setting custom fee rate provider")
+	}
+
+	// Collect weight and fee rate info about the sweep and new batch.
+	sweepFeeDetails, newBatchFeeDetails, err := estimateSweepFeeIncrement(
+		sweep,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to estimate tx weight for "+
+			"sweep %x: %w", sweep.swapHash[:6], err)
+	}
+
+	// Collect weight and fee rate info about existing batches.
+	batches := make([]feeDetails, 0, len(b.batches))
+	for _, existingBatch := range b.batches {
+		batchFeeDetails, err := estimateBatchWeight(existingBatch)
+		if err != nil {
+			return fmt.Errorf("failed to estimate tx weight for "+
+				"batch %d: %w", existingBatch.id, err)
+		}
+		batches = append(batches, batchFeeDetails)
+	}
+
+	// Run the algorithm. Get batchId of the best batch, or newBatchSignal
+	// if the best option is to create new batch.
+	batchId, err := selectBatch(
+		batches, sweepFeeDetails, newBatchFeeDetails,
+	)
+	if err != nil {
+		return fmt.Errorf("batch selection algorithm failed for sweep "+
+			"%x: %w", sweep.swapHash[:6], err)
+	}
+
+	// If the best option is to start new batch, do it.
+	if batchId == newBatchSignal {
+		return b.spinUpNewBatch(ctx, sweep)
+	}
+
+	// Locate the batch to add the sweep to.
+	bestBatch, has := b.batches[batchId]
+	if !has {
+		return fmt.Errorf("batch selection algorithm returned "+
+			"batch id %d which doesn't exist, for sweep %x",
+			batchId, sweep.swapHash[:6])
+	}
+
+	// Add the sweep to the batch.
+	accepted, err := bestBatch.addSweep(ctx, sweep)
```

Comment:
This may need to be enhanced

Currently, `addSweep` will also run checks that are related to the timeout window of the batch vs the sweep that is being added. See here https://github.com/lightninglabs/loop/blob/master/sweepbatcher/sweep_batch.go#L440-L447

This means that the single best solution that the greedy algorithm has calculated may be rejected because the timeout heights of the sweeps are too far away. I believe this is not the desired behavior?

I see 2 possible paths here:

a) Create a version of `addSweep` that does not check for the timeout window limits, allowing swaps with very distant timeout heights to be in the same batch
b) Enhance algorithm to return an array of "best batches", sorted from best to worst. Then on the above line (`76`) we would have to iterate over the results and try to `addSweep` in order from the best to worst option

Reply:
How often does it happen that a sweep is rejected based on the timeout? If that's not a common case then we can KISS and just try and add to the selected greedy batch, and if timeout prevents addition then fail the greedy algo and fall back to the default. Wdyt?

Reply:
> Didn't see your c), is a high failure rate expected?

That highly depends on the type of looper, frequent loopers will probably encounter this more often, as they are the ones having many batches running in parallel

Reply:
I like the option b) Enhance algorithm to return an array of "best batches", sorted from best to worst.

The selection algorithm already has a metric by which is compares all the alternatives, so it is easy to extend to return a sorted array.

Reply:
I implemented it and pushed a temporary commit with this change. Please take a look! I'll squash it before merging.